Fix concurrency issue in TypeDescriptor GetProperties() and GetConverter()

steveharter opened the pull request:
Fixes #92394
Fixes #30024

Replaces (and is derived from the same branch as) PR https://github.com/dotnet/runtime/pull/92521 in order to apply the "sentinel" pattern so that we don't need a new Hashtable instance. Thanks [karakasa](https://github.com/karakasa) and [Maximys](https://github.com/Maximys) for your work on this.

karakasa committed:
minimal entirety lock to fix bug

karakasa committed:
remove unused code

karakasa committed:
minimize lock range

karakasa committed:
add test

karakasa committed:
fix code style

karakasa committed:
Make `TypeDescriptor` thread safe with custom providers

Originally a race condition exists in `CheckDefaultProvider` and leads
to wrong results when many methods are called simultaneously.

The PR fixes that by extending the lock statement.

Fix #92934

karakasa committed:
Originally a race condition exists in `CheckDefaultProvider` and leads
to wrong results when many methods are called simultaneously.

The PR fixes that by extending the lock statement.

Fix #92394

karakasa committed:
Added another HashTable to reduce locks.

Moved tests to the main testing file
Adopted tests from #85156

Co-authored-by: Maximys <mixim33@yandex.ru>

karakasa committed:
better test names

karakasa committed:
improve comments

karakasa committed:
renamed variable

karakasa committed:
reverted sln file

karakasa committed:
Skip one test on browsers

ConcurrentGetProperties_ReturnsExpected
is skipped on browsers because Thread.Start
is unsupported.

steveharter committed:
Merge remote-tracking branch 'upstream/main' into issue-92394-v2

steveharter committed:
Change to sentinel approach

steveharter committed:
Merge branch 'main' into issue-92394-v2

ericstj commented on src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs:
```diff
@@ -271,33 +274,35 @@ private static void CheckDefaultProvider(Type type)
+                    }
                 }
+
+                s_defaultProviders[type] = s_initializedProvider;
```
I _think_ this needs to happen after the call to `CheckDefaultProvider(baseType);` and that needs to remain under the lock.  Otherwise another thread might exit out of this method before the default provider is added for the base type.

To test that you could try testing this with a race where the converter is not set on the target type, but it's base type.

steveharter replied:
Agree - expanded the lock. Also for the base type check, split the method into 2 to prevent an unnecessary dictionary check and a recursive lock.

ericstj replied:
Makes sense - that definitely makes this clearer

ericstj commented on src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs:
```diff
@@ -271,33 +274,35 @@ private static void CheckDefaultProvider(Type type)
 
-                // Immediately clear this. If we find a default provider
-                // and it starts messing around with type information,
+                // Immediately set this so we can short circuit in the ContainsKey check above.
+                // If we find a default provider and it starts messing around with type information,
```
I still don't understand this comment.  This is how I would word it.
```suggestion
                // Immediately set this to null to indicate we are in progress setting the default provider for a type.
                // This prevents re-entrance in this method.
```

steveharter replied:
Thanks - added

ericstj reviewed (commented):
This looks good, just one functional suggestion.

steveharter committed:
Expand lock for base type check; misc naming and comments

steveharter commented on src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs:
```diff
@@ -28,7 +28,8 @@ public sealed class TypeDescriptor
         private static readonly WeakHashtable s_providerTable = new WeakHashtable();     // mapping of type or object hash to a provider list
         private static readonly Hashtable s_providerTypeTable = new Hashtable();         // A direct mapping from type to provider.
-        private static readonly Hashtable s_defaultProviders = new Hashtable(); // A table of type -> default provider to track DefaultTypeDescriptionProviderAttributes.
+        private static readonly Hashtable s_defaultProviders = new Hashtable();          // A table of type -> default provider to track DefaultTypeDescriptionProviderAttributes.
```
Thanks - done

steveharter committed:
Merge remote-tracking branch 'upstream/main' into issue-92394-v2

steveharter committed:
Merge branch 'issue-92394-v2' of https://github.com/steveharter/runtime into issue-92394-v2

ericstj reviewed (approved):
This looks good to me.   For those tests - assume you saw them failing before the fix and confirmed the fix fixes them?

steveharter commented:
> This looks good to me. For those tests - assume you saw them failing before the fix and confirmed the fix fixes them?

Yes, I reverted the changes from the previous PR and the 3 new tests failed. Then I added back the changes and the tests passed.

cincuranet commented:
Regressions:
- https://github.com/dotnet/perf-autofiling-issues/issues/27477
